src/wallet/spend.cpp

```diff
@@ -1018,14 +1018,28 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
 
     coinControl.fAllowOtherInputs = true;
 
-    for (const CTxIn& txin : tx.vin) {
-        coinControl.Select(txin.prevout);
-    }
-
     // Acquire the locks to prevent races to the new locked unspents between the
     // CreateTransaction call and LockCoin calls (when lockUnspents is true).
     LOCK(wallet.cs_wallet);
 
+    // Fetch specified UTXOs from the UTXO set to get the scriptPubKeys and values of the outputs being selected
+    // and to match with the given solving_data. Only used for non-wallet outputs.
+    std::map<COutPoint, Coin> coins;
+    for (const CTxIn& txin : tx.vin) {
+        coins[txin.prevout]; // Create empty map entry keyed by prevout.
+    }
+    wallet.chain().findCoins(coins);
```

Comment:
At this point, we have `cs_wallet` locked (lock is few lines above) and inside `findCoins` we lock `cs_main`

Wonder if it's a potential deadlock cause or not. Because, `cs_main` should be taking precedence over `cs_wallet` in other processes.

Reply:
Agree it could be a problem in theory, but I'd like to understand the problem better before making a change.

Right now I couldn't find any places in our code that acquire `cs_wallet` while holding `cs_main`. To experiment I added `LOCK(cs_main)` to `wallet/rpc/coins.cpp:197` and immediately this triggered our deadlock detection mechanism because it turns out that `loadwallet` rpc acquires lock in a different order (first `cs_wallet` and then `cs_main`).

What is our long-term direction with regards to using `cs_main` and `cs_wallet` together? 
@achow101 any thoughts?

I'm also curious how we should think of these locks in light of multiprocess (cc @ryanofsky )

Reply:
`cs_main` is really for things in node operation and so there should never be a scenario within the wallet where we lock `cs_main` before `cs_wallet`. That would imply that the node is doing something to the wallet, but the model we (want to) have is the wallet asks for things from the node and otherwise operates asynchronously from the node. So it should always be `cs_wallet` before `cs_main`.

Reply:
@furszy Looks like the current lock order is fine. Is there anything left before you can ACK this?